neo/Consensus/ConsensusService.cs

```diff
@@ -225,18 +273,267 @@ private void OnPersistCompleted(Block block)
+            tempContext.Reset(0, context.Snapshot);
+            tempContext.Reset(message.ViewNumber, context.Snapshot);
+            if (message.PrepareRequestMessage != null)
+            {
+                var prepareRequest = message.PrepareRequestMessage;
+                tempContext.Nonce = prepareRequest.Nonce;
+                tempContext.TransactionHashes = prepareRequest.TransactionHashes;
+                tempContext.NextConsensus = prepareRequest.NextConsensus;
+                tempContext.Transactions = new Dictionary<UInt256, Transaction>
+                {
+                    [prepareRequest.TransactionHashes[0]] = prepareRequest.MinerTransaction
+                };
+                tempContext.Timestamp = prepareRequest.Timestamp;
+            }
+
+            ConsensusPayload prepareRequestPayload = null;
+            PrepareRequest prepareRequestMessage = null;
+
+            UInt256 preparationHash;
+            if (message.PreparationMessages.ContainsKey((int)tempContext.PrimaryIndex)
+                && ReverifyPrepareRequest(tempContext, message, out prepareRequestPayload, out prepareRequestMessage))
+                preparationHash = prepareRequestPayload.Hash;
+            else
+                preparationHash = message.PreparationHash;
+
+            var prepareResponses = new List<(ConsensusPayload, PrepareResponse)>();
+            var prepareResponseMsg = new PrepareResponse
+            {
+                ViewNumber = tempContext.ViewNumber,
+                PreparationHash = preparationHash
+            };
+            bool canRestoreView = false;
+            if (preparationHash != null)
+            {
+                for (ushort i = 0; i < context.Validators.Length; i++)
+                {
+                    if (i == tempContext.PrimaryIndex) continue;
+                    if (!message.PreparationMessages.ContainsKey(i)) continue;
+                    Log($" considering prepare request {i}");
+                    var regeneratedPrepareResponse = tempContext.RegenerateSignedPayload(prepareResponseMsg, i,
+                        message.PreparationMessages[i].InvocationScript);
+                    if (!regeneratedPrepareResponse.Verify(context.Snapshot) ||
+                        !PerformBasicConsensusPayloadPreChecks(regeneratedPrepareResponse)) continue;
+                    prepareResponses.Add((regeneratedPrepareResponse, prepareResponseMsg));
+                    Log($" verified prepare {i}");
+                    // Verify that there are M valid preparations, 1 Prepare Request + (M-1) Prepare responses
+                    if (prepareRequestMessage == null || prepareResponses.Count < context.M - 1) continue;
+                    canRestoreView = true;
+                    break;
+                }
+            }
+
+            // Only accept recovery from lower views if there were at least M valid prepare requests
+            if (message.ViewNumber < context.ViewNumber && !canRestoreView) return;
```

Comment:
If a node is in a higher view, it means that at least `M` nodes have sent requests to change the view to that number, and most likely they are all in the same view. In this case, even if there are `M` valid `Preparation`s in the lower view, they are unlikely to send `M` `Commit`s in the lower view. In this case, it is better to disallow from changing to a lower view and let them reach a consensus in a higher view. Otherwise it is possible for a node to constantly switch back and forth between the lower view and the higher view.

Reply:
If we have `M` valid Preparations that node will jump to a lower view and automatically set `CommitFlag` and, thus, will stuck in such commit state.

Reply:
Without such mechanism, only 2 faulty nodes could stuck the network.
- Let's suppose that a  given node (n1) reach `commit` (`M` preparations) and recover `f-1` nodes when they asks for `change_view`;
- The remainder `M` nodes change view;
- If any of those `M` nodes fail (nX) they will never reach a consensus (n1 and nX were the only byzantine nodes, while the other one recovered did not failed, they were just recovered exactly at the right time).

On the other hand, such mechanism is not perfect. The problem we saw was a split when we have several commitments achieved in different views. The trade-off is hard to ponder, Erik.
What do you think?
Personally, I think that after a commit is found that node should spread that good news and try to get "followers".

Reply:
Let us consider a network with 4 consensus nodes, A, B, C and D.

When view number is `0`, A, B, and C sent the `Preparation`s, but for some reason they did not reach a consensus, so they entered the view `1`. But D is still in view `0`.

When view number is `1`, A and B sent the `Commit`s, C crashed. At the same time, D received the `Preparation`s in view `0` and sent `Commit`.

After C restarts, it will receive `RecoveryMessage`s from A, B, and D. These recovery messages contain `M` valid `Preparation`s. But obviously the probability of reaching a consensus in view `1` is higher.

So I think it is necessary to only allow moving to higher views.

Reply:
> If any of those M nodes fail (nX) 

This shows that the number of fault nodes has exceeded `f`. In this case, no mechanism can achieve a reliable consensus.

Reply:
Switching to a higher view means that most nodes already believe that the old view's messages are all invalid, so you should not continue to use old messages under any circumstances.

Reply:
@erikzhang I tend to agree with you, and this is how I had it initially. However, @vncoelho seemed convinced that some could count to the next. Thinking over it again, I don't think that can actually happen, and we should just only allow moving the view forward.

Reply:
Actually, let me try to propose a case why this is useful still and we can discuss.

Reply:
Here is a situation where it can still be helpful:
Start in view 0
A -> crashed
B -> send prepare request
C -> send prepare response
D -> send prepare reponse
receive C prepare response -> B
receive D prepare -> B
A -> return from crash around the same time B loses network connection
B -> send commit (but now lost network connection, no one recieves)
-------------------------------------
C & D timers timeout, and they decide to send change view 1
A -> timeout and agrees to change view to view 1
A, B, and C now in view 1
B -> gets network connection back (stuck in commit in view 0)
C -> crash
Now, B sends recovery periodically since he is committed; if A, and C are allowed to recover back to view 0, then they will immediately accept the message and all will commit in view 0 and block height will continue.
----------------------------------------

@erikzhang  The drawback you are discussing about potentially changing view back and forth is what I brought up earlier in this PR with @vncoelho ; however, since once committed they can't change view, and the criteria for going back allows them to commit immediately they won't go back and forth; however, they can get stuck in a split scenario with this code that allows them to go back, but they can get stuck in a split scenario anyway with out this commit that allows them to go back if more than `f` nodes fail. In the scenario given above there are never more than 1 node really completely failed at any given time and it proposes why the current solution is better in that situation.

Honestly, I could accept it either way, but I think the current solution in practice will be a bit better, and will likely not if ever stall the network in a split commit situation. If it does ever happen, all that will be necessary is for nodes on one of the views to delete their consensus storage and restart to get things back flowing. I'm not sure that will ever happen in practice even. It isn't even possible with 4 consensus nodes I think; with 7 it is pretty complicated to write the scenario where it happens but I have it written out on this PR earlier already once.

Reply:
Please wait and review the next change I add to this PR to fix the proper criteria for allowing the view to go backward to join an already committed node in a lower view.

Reply:
Ok, I fixed the criteria for jumping to a prior view; nodes have to be joining a committed node to move their view backward.

Reply:
@erikzhang Is this acceptable to you now?

Reply:
In fact, our algorithm is based on the theory of pbft. There is no mention in pbft that the view can be rolled back. If we want to do this, I think it is necessary to give a theoretical proof that it is safe, otherwise it should not take this risk.

Reply:
So I highly recommend that we currently only allow the view to be moved forward.

Reply:
@erikzhang that is fair to me. There is no reason to allow it to move backward now in my opinion. I mainly added it because essentially the above scenario was something that @vncoelho was indicating to me was somewhat of a requirement for what he had in mind.

Reply:
Since the code allowing it is all basically in one spot, I'm going to comment it out with an explanation so we can merge without it, but do the research after.

Reply:
Maybe removing the code directly would be better. I will make some changes later.

Reply:
Ok, I will just remove it.

Reply:
I will fix the places that send the message that would make the view move backward then.

Reply:
I agree with you, @erikzhang.

@jsolman, let's move that mechanism back, sorry for that, and try without rolling back.
If needed, in the future we check such idea.

Both cases have their open weakness.